src/VecSim/algorithms/hnsw/hnsw_batch_iterator.cpp

```diff
@@ -0,0 +1,188 @@
+#include "hnsw_batch_iterator.h"
+
+#include <utility>
+#include "VecSim/query_result_struct.h"
+#include "VecSim/utils/vec_utils.h"
+#include "VecSim/algorithms/hnsw/visited_nodes_handler.h"
+
+inline void HNSW_BatchIterator::visitNode(idType node_id) {
+    this->visited_list->tagNode(node_id, this->visited_tag);
+}
+
+inline bool HNSW_BatchIterator::hasVisitedNode(idType node_id) const {
+    return this->visited_list->getNodeTag(node_id) == this->visited_tag;
+}
+
+VecSimQueryResult_List HNSW_BatchIterator::prepareResults(candidatesMaxHeap top_candidates,
+                                                          size_t n_res) {
+    // size_t initial_results_num = array_len(batch_results);
+    // Put the "spare" results (if exist) in the results heap.
+    while (top_candidates.size() > n_res) {
+        this->top_candidates_extras.emplace(top_candidates.top().first,
+                                            top_candidates.top().second); // (distance, id)
+        top_candidates.pop();
+    }
+    auto *batch_results =
+        array_new_len<VecSimQueryResult>(top_candidates.size(), top_candidates.size());
+    // Return results from the top candidates heap, put them in reverse order in the batch results
+    // array.
+    for (int i = (int)(top_candidates.size() - 1); i >= 0; i--) {
+        labelType label = this->hnsw_index->getExternalLabel(top_candidates.top().second);
+        VecSimQueryResult_SetId(batch_results[i], label);
+        VecSimQueryResult_SetScore(batch_results[i], top_candidates.top().first);
+        top_candidates.pop();
+    }
+    return batch_results;
+}
+
+candidatesMaxHeap HNSW_BatchIterator::scanGraph(candidatesMinHeap &candidates,
+                                                candidatesMinHeap &top_candidates_extras,
+                                                float &lower_bound, idType entry_point) {
+
+    candidatesMaxHeap top_candidates(this->allocator);
+    if (entry_point == -1) {
+        this->depleted = true;
+        return top_candidates;
+    }
+
+    // Move extras from previous iteration to the top candidates.
+    while (top_candidates.size() < this->hnsw_index->getEf() && !top_candidates_extras.empty()) {
+        top_candidates.emplace(top_candidates_extras.top().first,
+                               top_candidates_extras.top().second);
+        top_candidates_extras.pop();
+    }
+    if (top_candidates.size() == this->hnsw_index->getEf()) {
+        return top_candidates;
+    }
+
+    auto dist_func = this->space->get_dist_func();
+
+    // In the first iteration, add the entry point to the empty candidates set.
```

Comment:
Code arraignment - I think this should be right after the first condition (no entry point)